tools/developer/lps2jani/lps2jani.cpp

```diff
@@ -0,0 +1,452 @@
+#include "mcrl2/data/pos.h"
+#include "mcrl2/data/nat.h"
+#include "mcrl2/data/int.h"
+#include "mcrl2/data/real.h"
+#include "mcrl2/lps/io.h"
+#include "mcrl2/lps/is_stochastic.h"
+#include "mcrl2/lps/untime.h" // For real_zero. Should be moved to another file. TODO.
+
+#include <boost/json/src.hpp>
+
+using namespace mcrl2;
+using namespace boost::json;
+
+using enumerator_element = data::enumerator_list_element_with_substitution<>;
+using super = mcrl2::data::tools::rewriter_tool<mcrl2::utilities::tools::input_output_tool>;
+
+class data_structures
+{
+  data::rewriter m_rewr;
+  data::enumerator_identifier_generator m_id_generator;
+  data::enumerator_algorithm<> m_enumerator;
+  data::mutable_indexed_substitution<> m_sigma;
+
+public:
+  explicit data_structures(const lps::stochastic_specification& stochastic_lpsspec)
+    : m_rewr(stochastic_lpsspec.data()),
+    m_enumerator(m_rewr, stochastic_lpsspec.data(), m_rewr, m_id_generator, false)
+  {
+  }
+
+  data::data_expression rewriter(const data::data_expression& e) const
+  {
+    return m_rewr(e, m_sigma);
+  }
+
+  data::enumerator_algorithm<>& enumerator()
+  {
+    return m_enumerator;
+  }
+
+  data::mutable_indexed_substitution<>& sigma()
+  {
+    return m_sigma;
+  }
+};
+
+class lps2jani_tool : public super
+{
+  std::unique_ptr<data_structures> m_data_structures;
+public:
+  lps2jani_tool()
+    : super("lps2jani",
+      "Jan Friso Groote, PHM van Spaendonck",
+      "generates a Jani file from a (stochastic) LPS",
+      "Reads the LPS in INFILE and writes a corresponding file in Jani format "
+      "to OUTFILE. The LPS must only contain simple data types (Numbers, Booleans) that the "
+      "Jani format supports. See http://jani-spec.org/#Model for more information about the Jani format. "
+      "If OUTFILE is not present, stdout is used. If INFILE is not present, stdin is used."
+    )
+  {}
+
+private:
+  /**
+   * \brief Converts the given action label list to a JSON array of actions.
+   * Actions use the schema:
+   *  {"name": Identifier}
+   *
+   * \param act_decls the to-be converted action declaration list.
+   * \return A JSON array corresponding to \ref act_decls.
+   */
+  array convert_action_declarations(const process::action_label_list& act_decls) const
+  {
+    array actions;
+    for (const process::action_label& d : act_decls)
+    {
+      actions.push_back(object{ {"name", (std::string)d.name()} });
+    }
+    return actions;
+  }
+
+  /**
+   * \brief Converts the given OpId into the JSON string that corresponds to it according to the JANI schema.
+   *
+   * \param opid the to-be converted OpId.
+   * \return A JSON string corresponding to the operator.
+   */
+  string convert_operator_to_jani(const data::data_expression& opid) const {
+    if (std::string op = mcrl2::data::pp(opid);
+      "!=" == op) {
+      return "≠";
+    }
+    else if ("==" == op) {
+      return "=";
+    }
+    else if ("<=" == op) {
+      return "≤";
```

Comment:
None of this code is tested so I don't necessarily know what happens here, but ≤ is not an ASCII character and so I think should be prefixed by `u8` to indicate that it is a UTF-8 string literal. However, C++ source code should also be only within the so called basic character set so maybe it's even necessary to write some garbage like u8"\U+2264". Not entirely sure, UTF support is just garbage in C++ but compilers seem to support it sensibly.